### PR TITLE
123759 synthetics custom tags

### PR DESCRIPTION
**Pull request by kmcormick**
Add information for custom tags to Synthetic Test Playback Results descriptions

### CODE REVIEW

File: spec/descriptions/tagSyntheticResults.md [hunk 1]
```diff
@@ -20,8 +20,12 @@ The endpoint returns the aggregated Synthetic test result data
    response_time (ms), response_size (bytes), status_code (an integer represents an HTTP response code, e.g., 200, 401, 500), request_size (bytes), 
    upload_speed (bytes per second), download_speed (bytes per second), 
    redirect_time (ms), redirect_count, connect_count, and status (an integer, 1-success or 0-failure). 
+   
    The following metrics are only available for the HTTPAction type Synthetic Tests: blocking (bytes), dns (bytes), connect (bytes), ssl (bytes), 
    sending (bytes), waiting (bytes), and receiving (bytes).
+
+   The metric synthetic.tags will add the latest list of custom properties to the response.  This metric cannot be *aggregated*.
```
Comment (KarenGdaniec) [inline spec/descriptions/tagSyntheticResults.md:27]:
```suggestion
   The metric synthetic.tags adds the latest list of custom properties to the response.  This metric cannot be *aggregated*.
```

**Review by KarenGdaniec — CHANGES_REQUESTED** [review]
@kmcormick There is a paragraph at the top of the file that needs to be updated to list synthetic.tags as a metric.  I was not able to make a comment on that part of the file.  Copied the section to be updated into comment.

> **Note on names in TagFilter/TagFilterExpression**: From R243, the name used in a TagFilter or a TagFilterExpression has the format: synthetic.\<name\>.
> It can be one of the following: synthetic.id (id is the test result id), synthetic.testId,
> synthetic.testName, synthetic.locationId, synthetic.applicationId, synthetic.serviceId, synthetic.syntheticType,
> synthetic.locationName, and synthetic.locationLabel. If it is a metric name, then the format is: synthetic.metrics\<MetricName\>.
> For example, synthetic.metricsResponseTime, synthetic.metricsStatus.

**Comment by kmcormick** [comment]
@KarenGdaniec Yes, for the API call that returns "aggregated" results, when synthetic.tags is specified as a metric, then the latest properties are returned.  The sentence was meant to mean that synthetic.tags cannot be aggregated (i.e. you cannot get the sum of the value of key2 by specifying "aggregation":"SUM").

**Comment by KarenGdaniec** [comment]
@kmcormick okay, when I look at the whole doc, I didn't see your code example in each call, so I wasn't sure if it could be specified in that one.  This doc is a little odd to me with all the different calls, just check to make sure your example appears in each section where adding `synthetic.tags` is applicable.

**Comment by kmcormick** [comment]
@KarenGdaniec That is because the aggregated call does not list tagFilter and tagFilterExpression as optional.  I will check with Carol to see if that should be added.

**Comment by carolzhang65** [comment]
> @KarenGdaniec That is because the aggregated call does not list tagFilter and tagFilterExpression as optional. I will check with Carol to see if that should be added.

@kmcormick For "Get Synthetic test playback results" (aggregated) part, I did not list tagFilter or tagFilterExpression as an optional parameter. You can add them the same way as that in "Get a list of Synthetic test playback results (no aggregation)". Thanks.

File: spec/descriptions/tagSyntheticResults.md [hunk 2]
```diff
@@ -83,17 +101,17 @@ To narrow down the result set you have two options to search for a test.
 {
     "testId": ["tUmWgvzdo1Q1vpVRpzR5", "Pg0Q1UqHRd7OMysohVLd"],
     "//comment1": "Get test results from last 30 minutes (windowSize), data are aggregated every 10 minutes (granularity)",
-    "//comment2": "The granularity values for response_time and response_size must be the same"
+    "//comment2": "The granularity values for responsTime and responseSize must be the same"
```
Comment (KarenGdaniec) [inline spec/descriptions/tagSyntheticResults.md:104]:
```suggestion
    "//comment2": "The granularity values for responseTime and responseSize must be the same"
```

Comment (kmcormick) [reply spec/descriptions/tagSyntheticResults.md:104]: good eyes, thank you

File: spec/descriptions/tagSyntheticResults.md [hunk 3]
```diff
@@ -127,7 +145,7 @@ The metric synthetic.tags will add the latest list of custom properties to the r
 2. *pageSize* set the number of Synthetic test results you want to return with one query
 
 **order** You can order the returned items alphanumerical by label, either ascending or descending
-1. *by* Use the metric name, e.g. "response_time" to order by that value
+1. *by* Use the metric name, e.g. "synthetic.metricsResponseTime" to order by that value
```
[the remark this answers is not in the document]
Comment (kmcormick) [reply spec/descriptions/tagSyntheticResults.md:148]: @KarenGdaniec Did Carol confirm this is okay or is there a different issue that required orderBy to use the old names?

File: spec/descriptions/tagSyntheticResults.md [hunk 4]
```diff
@@ -79,17 +101,17 @@ To narrow down the result set you have two options to search for a test.
...
     "testId": ["tUmWgvzdo1Q1vpVRpzR5", "Pg0Q1UqHRd7OMysohVLd"],
     "//comment1": "Get test results from last 30 minutes (windowSize), data are aggregated every 10 minutes (granularity)",
-    "//comment2": "The granularity values for response_time and response_size must be the same"
+    "//comment2": "The granularity values for responseTime and responseSize must be the same"
     "metrics": [
      {
         "aggregation": "MEAN",
         "granularity": 600,    
-        "metric": "response_time"
+        "metric": "synthetic.metricsResponseTime"
```
Comment (carolzhang65) [inline spec/descriptions/tagSyntheticResults.md:109]: Both "response_time" and "synthetic.metricsResponseTime" work here. Same for the response_size and synthetic.metricsResponseSize.

File: spec/descriptions/tagSyntheticResults.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (carolzhang65) [reply spec/descriptions/tagSyntheticResults.md:148]: Adding:
```
"order": {
     	"by": "synthetic.metricsResponseTime",
     	"direction": "DESC"
     }
```
got internal server error. I have not checked logs in Pink.

File: spec/descriptions/tagSyntheticResults.md [hunk 5]
```diff
@@ -129,14 +154,24 @@ It will be replaced by **tagFilterExpression**.
...
+  "value":"Denver",
+  "operator":"EQUALS"
+}]
+```
 
 ### Sample payload to get a list of Synthetic test results with tagFilters
 ```json
 {
-  "syntheticMetrics":["response_time","response_size"],
+  "syntheticMetrics":["synthetic.metricsResponseTime","synthetic.metricsResponseSize"],
```
Comment (carolzhang65) [inline spec/descriptions/tagSyntheticResults.md:172]: For syntheticMetrics, both response_time and synthetic.metricsResponseTime work.

File: spec/descriptions/tagSyntheticResults.md [hunk 6]
```diff
@@ -129,14 +154,24 @@ It will be replaced by **tagFilterExpression**.
...
   "order":{
-    "by":"response_time",
+    "by":"synthetic.metricsResponseTime",
```
Comment (carolzhang65) [inline spec/descriptions/tagSyntheticResults.md:174]: Both response_time and synthetic.metricsResponseTime work fine.

File: spec/descriptions/tagSyntheticResults.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmcormick) [reply spec/descriptions/tagSyntheticResults.md:148]: Was that for getting test cases or testSummaryList?  I tried the following for the former and it worked:
`    OrderByDTO orderByASC = OrderByDTO.newBuilder().setByLabel("synthetic.metricsResponseTime").setDirection("ASC").build();
    TestResultsDTO result = getTestResultsCommandHandler.getTestResults(TestResultQueryDTO.newBuilder()
            .setClientId(clientId)
            .addTestId(TEST_ID_1_VALUE)
            .addMetricParameters(TestResultsMetricParameterDTO.newBuilder()
                .setMetric("response_size"))
            .setTimeConfig(TimeConfigDTO.newBuilder()
                .setWindowSize(windowSize)
                .setTo(to))
            .setOrderBy(orderByASC)
            .build())
        .block();`

File: spec/descriptions/tagSyntheticResults.md [hunk 7]
```diff
@@ -48,9 +52,27 @@ The timeFrame might be adjusted to fit the metric granularity so that there is n
 2. *pageSize* set the number of Synthetic test results you want to return with one query
 
 **order** You can order the returned items alphanumerical by label, either ascending or descending
-1. *by* Use the metric name, e.g. "response_time", to order by its value
+1. *by* Use the metric name, e.g. "synthetic.metricsResponseTime", to order by its value
```
Comment (carolzhang65) [inline spec/descriptions/tagSyntheticResults.md:55]: For aggregated test results, adding:
```
"order": {
     	"by": "synthetic.metricsResponseTime",
     	"direction": "DESC"
     }
```
got internal server error. This is the error in synthetics-reader in Pink: 
```
Caused by: com.instana.clickhouse.client.exception.ClickHouseRequestException: Code: 46. DB::Exception: Unknown function avgIf. (UNKNOWN_FUNCTION) (version 22.3.15.33 (official build))
. Query: SELECT client_id,test_id,toString((floor((start_time/600000))*600000)) AS `start_time_aggregated`,avgIf(metrics_response_time,metrics_response_time != 9223372036854775807) AS `metrics_response_time`,avgIf(metrics_status,metrics_status != 2147483647) AS `metrics_status`,metrics_response_time,groupUniqArray(location_id) FROM synthetics.all_synthetic_test_results LEFT JOIN locations_temp_table_6be1d810 ON (synthetics.all_synthetic_test_results.location_id = gk_location_id) WHERE client_id = 'saas_instana_test' AND start_time >= 1682525400000 AND start_time <= 1682532000000 AND test_id IN ('s8Cr2aTFMLUGufmd0vNK') GROUP BY client_id,test_id,start_time_aggregated,metrics_response_time,start_time_aggregated ORDER BY metrics_response_time DESC,start_time_aggregated ASC LIMIT 100 FORMAT JSON
```

File: spec/descriptions/tagSyntheticResults.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (carolzhang65) [reply spec/descriptions/tagSyntheticResults.md:148]: For list of test results, both response_time and synthetic.metricsResponseTime work fine.

File: spec/descriptions/tagSyntheticResults.md (hunk 7 above)
Comment (carolzhang65) [reply spec/descriptions/tagSyntheticResults.md]: Without order by response_time or order by synthetic.metricsResponseTime, the following payload works fine: 
{
    "testId": ["s8Cr2aTFMLUGufmd0vNK"],
    "//comment1": "Get test results from last 30 minutes (windowSize), data are aggregated every 10 minutes (granularity)",
    "//comment2": "status = 1 ==> successful",
    "metrics": [
     {
        "aggregation": "MEAN",
        "granularity": 600,
        "metric": "synthetic.metricsResponseTime"
     },
     {
        "aggregation": "MEAN",
        "granularity": 600,
        "metric": "status"
     }],
     "timeFrame": {
       "to": 0,
       "windowSize": 7200000
     },
     "tagFilters":[{
     	"name": "synthetic_type",
     	"operation": "EQUALS",
     	"stringValue": "BrowserScript"
     }]
}

Comment (kmcormick) [reply spec/descriptions/tagSyntheticResults.md]: I created a card to investigate further:  https://instana.kanbanize.com/ctrl_board/132/cards/125989/details/

**Review by carolzhang65 — APPROVED** [review]
The changes look good to me. Thanks.

**Pull request merged**